master/buildbot/reporters/telegram.py

```diff
@@ -0,0 +1,1055 @@
+# This file is part of Buildbot.  Buildbot is free software: you can
+# redistribute it and/or modify it under the terms of the GNU General Public
+# License as published by the Free Software Foundation, version 2.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+
+import io
+import json
+import random
+import shlex
+
+from twisted.internet import defer
+from twisted.internet import task
+from twisted.python import log
+from twisted.web import resource
+from twisted.web import server
+
+from buildbot import config
+from buildbot import util
+from buildbot.plugins.db import get_plugins
+from buildbot.process.results import CANCELLED
+from buildbot.process.results import EXCEPTION
+from buildbot.process.results import FAILURE
+from buildbot.process.results import RETRY
+from buildbot.process.results import SUCCESS
+from buildbot.process.results import WARNINGS
+from buildbot.reporters.words import Channel
+from buildbot.reporters.words import Contact
+from buildbot.reporters.words import StatusBot
+from buildbot.reporters.words import UsageError
+from buildbot.schedulers.forcesched import CollectedValidationError
+from buildbot.schedulers.forcesched import ForceScheduler
+from buildbot.util import bytes2unicode
+from buildbot.util import httpclientservice
+from buildbot.util import service
+from buildbot.util import unicode2bytes
+
+
+class TelegramChannel(Channel):
+
+    def __init__(self, bot, channel):
```

Comment:
Could we pass channel info and id as separate arguments? The current approach will make code harder to understand and maintain in the future.

What I would propose is having `channel_info` and `channel_id`. Both are `None` by default and exactly one is accepted. `channel_id` would be inferred from `channel_info` dict if it is specified. `channel_id` would be sent to `super()`. `channel_info` would be stored as `self.chat`.

Reply:
I strongly prefer not to change the signature of the overridden method (including constructor) if not absolutely necessary.

I have changed the object attribute names (to `chat_info`) and refactored the code so `TelegramChannel` constructor **always** gets a dict.

Reply:
Thanks, sounds reasonable.